### ISSUE TITLE
Cannot mock class with constructor arguments of generic type

### ISSUE THREAD

**Issue by HylkeB** [lupuuss/Mokkery#66]
I cannot create a mock for an open class (using allopen plugin) when the class has a constructor that accepts a generic type.

Would be nice if there is a solution for this.

See reproducer code below:

```kotlin
@OpenForMocking
class ClassWithGenericsInConstructor<R>(
    val value: R,
) {
    fun doStuff(): R {
        return value
    }
}

@OpenForMocking
class ClassWithoutGenericsInConstructor<R>(
    val value: String,
) {
    fun doStuff(thing: R) {
        println("ok")
    }
}

class TestClass {
    // The following line crashes
    private val classWithGenerics = mock<ClassWithGenericsInConstructor<String>>()

    // But this is all fine (if i comment out the line above)
    private val classWithoutGenerics = mock<ClassWithoutGenericsInConstructor<String>>()
    // I can even do all mocks, verifies etc with this instance.
}
```

The stacktrace  is:
```
org.jetbrains.kotlin.backend.common.BackendException: Backend Internal error: Exception during IR lowering
File being compiled: /path/to/my/test.kt
The root cause java.lang.RuntimeException was thrown at: org.jetbrains.kotlin.backend.jvm.codegen.FunctionCodegen.generate(FunctionCodegen.kt:47)
	at org.jetbrains.kotlin.backend.common.CodegenUtil.reportBackendException(CodegenUtil.kt:253)
	at org.jetbrains.kotlin.backend.common.CodegenUtil.reportBackendException$default(CodegenUtil.kt:236)
	at org.jetbrains.kotlin.backend.common.phaser.PerformByIrFilePhase.invokeSequential(performByIrFile.kt:65)
	at org.jetbrains.kotlin.backend.common.phaser.PerformByIrFilePhase.invoke(performByIrFile.kt:52)
	at org.jetbrains.kotlin.backend.common.phaser.PerformByIrFilePhase.invoke(performByIrFile.kt:38)
	at org.jetbrains.kotlin.backend.common.phaser.NamedCompilerPhase.phaseBody(CompilerPhase.kt:166)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedCompilerPhase.invoke(CompilerPhase.kt:113)
	at org.jetbrains.kotlin.backend.common.phaser.CompositePhase.invoke(PhaseBuilders.kt:27)
	at org.jetbrains.kotlin.backend.common.phaser.CompositePhase.invoke(PhaseBuilders.kt:14)
	at org.jetbrains.kotlin.backend.common.phaser.NamedCompilerPhase.phaseBody(CompilerPhase.kt:166)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedCompilerPhase.invoke(CompilerPhase.kt:113)
	at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.invokeToplevel(CompilerPhase.kt:62)
	at org.jetbrains.kotlin.backend.jvm.JvmIrCodegenFactory.invokeCodegen(JvmIrCodegenFactory.kt:371)
	at org.jetbrains.kotlin.codegen.CodegenFactory.generateModule(CodegenFactory.kt:47)
	at org.jetbrains.kotlin.backend.jvm.JvmIrCodegenFactory.generateModuleInFrontendIRMode(JvmIrCodegenFactory.kt:433)
	at org.jetbrains.kotlin.cli.jvm.compiler.pipeline.JvmCompilerPipelineKt.generateCodeFromIr(jvmCompilerPipeline.kt:246)
	at org.jetbrains.kotlin.cli.jvm.compiler.pipeline.JvmCompilerPipelineKt.compileModulesUsingFrontendIrAndLightTree(jvmCompilerPipeline.kt:142)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:148)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:43)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:103)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:49)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:101)
	at org.jetbrains.kotlin.incremental.IncrementalJvmCompilerRunner.runCompiler(IncrementalJvmCompilerRunner.kt:464)
	at org.jetbrains.kotlin.incremental.IncrementalJvmCompilerRunner.runCompiler(IncrementalJvmCompilerRunner.kt:73)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.doCompile(IncrementalCompilerRunner.kt:506)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileImpl(IncrementalCompilerRunner.kt:423)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.tryCompileIncrementally$lambda$9$compile(IncrementalCompilerRunner.kt:249)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.tryCompileIncrementally(IncrementalCompilerRunner.kt:267)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compile(IncrementalCompilerRunner.kt:120)
	at org.jetbrains.kotlin.daemon.CompileServiceImplBase.execIncrementalCompiler(CompileServiceImpl.kt:675)
	at org.jetbrains.kotlin.daemon.CompileServiceImplBase.access$execIncrementalCompiler(CompileServiceImpl.kt:92)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl.compile(CompileServiceImpl.kt:1660)
	at jdk.internal.reflect.GeneratedMethodAccessor13.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(Unknown Source)
	at java.rmi/sun.rmi.transport.Transport$1.run(Unknown Source)
	at java.rmi/sun.rmi.transport.Transport$1.run(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Unknown Source)
	at java.rmi/sun.rmi.transport.Transport.serviceCall(Unknown Source)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(Unknown Source)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(Unknown Source)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Unknown Source)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.RuntimeException: Exception while generating code for:
CONSTRUCTOR visibility:public <> (mode:dev.mokkery.MockMode, block:kotlin.Function1<kotlin.Any?, kotlin.Any?>?) returnType:com.example.ClassWithGenericsInConstructorMockff95e2a7ef5741dd9fe506074f570d58 [primary]
  VALUE_PARAMETER name:mode index:0 type:dev.mokkery.MockMode
  VALUE_PARAMETER name:block index:1 type:kotlin.Function1<kotlin.Any?, kotlin.Any?>?
  BLOCK_BODY
    DELEGATING_CONSTRUCTOR_CALL 'public constructor <init> (value: R of com.example.ClassWithGenericsInConstructor) [primary] declared in com.example.ClassWithGenericsInConstructor'
      <R>: <none>
      value: CALL 'internal final fun autofillConstructor <T> (type: kotlin.reflect.KClass<*>): T of dev.mokkery.internal.CompilerPluginUtilsKt.autofillConstructor declared in dev.mokkery.internal.CompilerPluginUtilsKt' type=R of com.example.ClassWithGenericsInConstructor origin=null
        <T>: R of com.example.ClassWithGenericsInConstructor
        type: CLASS_REFERENCE 'TYPE_PARAMETER name:R index:0 variance: superTypes:[kotlin.Any?] reified:false' type=kotlin.reflect.KClass<*>
    SET_FIELD 'FIELD PROPERTY_BACKING_FIELD name:interceptedTypes type:kotlin.collections.List<kotlin.reflect.KClass<*>> visibility:public' type=kotlin.Unit origin=null
      receiver: GET_VAR '<this>: com.example.ClassWithGenericsInConstructorMockff95e2a7ef5741dd9fe506074f570d58 declared in com.example.ClassWithGenericsInConstructorMockff95e2a7ef5741dd9fe506074f570d58' type=com.example.ClassWithGenericsInConstructorMockff95e2a7ef5741dd9fe506074f570d58 origin=null
      value: CALL 'public final fun listOf <T> (vararg elements: T of kotlin.collections.CollectionsKt.listOf): kotlin.collections.List<T of kotlin.collections.CollectionsKt.listOf> declared in kotlin.collections.CollectionsKt' type=kotlin.collections.List<T of kotlin.collections.CollectionsKt.listOf> origin=null
        <T>: <none>
        elements: BLOCK type=kotlin.Array<kotlin.reflect.KClass<T of kotlin.reflect.KClass>> origin=null
          VAR IR_TEMPORARY_VARIABLE name:tmp0 type:kotlin.Array<kotlin.reflect.KClass<T of kotlin.reflect.KClass>> [val]
            CALL 'public final fun arrayOfNulls <T> (size: kotlin.Int): kotlin.Array<T of kotlin.arrayOfNulls?> declared in kotlin' type=kotlin.Array<kotlin.reflect.KClass<T of kotlin.reflect.KClass>> origin=null
              <T>: kotlin.reflect.KClass<T of kotlin.reflect.KClass>
              size: CONST Int type=kotlin.Int value=1
          CALL 'public final fun set (index: kotlin.Int, value: T of kotlin.Array): kotlin.Unit [operator] declared in kotlin.Array' type=kotlin.Unit origin=null
            $this: GET_VAR 'val tmp0: kotlin.Array<kotlin.reflect.KClass<T of kotlin.reflect.KClass>> [val] declared in com.example.ClassWithGenericsInConstructorMockff95e2a7ef5741dd9fe506074f570d58.<init>' type=kotlin.Array<kotlin.reflect.KClass<T of kotlin.reflect.KClass>> origin=null
            index: CONST Int type=kotlin.Int value=0
            value: CLASS_REFERENCE 'CLASS CLASS name:ClassWithGenericsInConstructor modality:OPEN visibility:public superTypes:[kotlin.Any]' type=kotlin.reflect.KClass<*>
          GET_VAR 'val tmp0: kotlin.Array<kotlin.reflect.KClass<T of kotlin.reflect.KClass>> [val] declared in com.example.ClassWithGenericsInConstructorMockff95e2a7ef5741dd9fe506074f570d58.<init>' type=kotlin.Array<kotlin.reflect.KClass<T of kotlin.reflect.KClass>> origin=null
    SET_FIELD 'FIELD PROPERTY_BACKING_FIELD name:interceptor type:dev.mokkery.internal.MokkeryMock visibility:public' type=kotlin.Unit origin=null
      receiver: GET_VAR '<this>: com.example.ClassWithGenericsInConstructorMockff95e2a7ef5741dd9fe506074f570d58 declared in com.example.ClassWithGenericsInConstructorMockff95e2a7ef5741dd9fe506074f570d58' type=com.example.ClassWithGenericsInConstructorMockff95e2a7ef5741dd9fe506074f570d58 origin=null
      value: CALL 'internal final fun MokkeryMock (mockMode: dev.mokkery.MockMode, kind: dev.mokkery.internal.MokkeryKind): dev.mokkery.internal.MokkeryMock declared in dev.mokkery.internal.MokkeryMockKt' type=dev.mokkery.internal.MokkeryMock origin=null
        mockMode: GET_VAR 'mode: dev.mokkery.MockMode declared in com.example.ClassWithGenericsInConstructorMockff95e2a7ef5741dd9fe506074f570d58.<init>' type=dev.mokkery.MockMode origin=null
        kind: GET_FIELD 'FIELD FIELD_FOR_ENUM_ENTRY name:Mock type:dev.mokkery.internal.MokkeryKind visibility:public [final,static] declared in dev.mokkery.internal.MokkeryKind' type=dev.mokkery.internal.MokkeryKind origin=null
    SET_FIELD 'FIELD PROPERTY_BACKING_FIELD name:id type:kotlin.String visibility:public' type=kotlin.Unit origin=null
      receiver: GET_VAR '<this>: com.example.ClassWithGenericsInConstructorMockff95e2a7ef5741dd9fe506074f570d58 declared in com.example.ClassWithGenericsInConstructorMockff95e2a7ef5741dd9fe506074f570d58' type=com.example.ClassWithGenericsInConstructorMockff95e2a7ef5741dd9fe506074f570d58 origin=null
      value: CALL 'internal final fun generateMockId (typeName: kotlin.String): kotlin.String declared in dev.mokkery.internal.CompilerPluginUtilsKt' type=kotlin.String origin=null
        typeName: CONST String type=kotlin.String value="ClassWithGenericsInConstructor"
    WHEN type=kotlin.Unit origin=null
      BRANCH
        if: CALL 'public final fun not (): kotlin.Boolean [operator] declared in kotlin.Boolean' type=kotlin.Boolean origin=EXCL
          $this: CALL 'public final fun EQEQ (arg0: kotlin.Any?, arg1: kotlin.Any?): kotlin.Boolean declared in kotlin.internal.ir' type=kotlin.Boolean origin=EQEQ
            arg0: GET_VAR 'block: kotlin.Function1<kotlin.Any?, kotlin.Any?>? declared in com.example.ClassWithGenericsInConstructorMockff95e2a7ef5741dd9fe506074f570d58.<init>' type=kotlin.Function1<kotlin.Any?, kotlin.Any?>? origin=null
            arg1: CONST Null type=kotlin.Nothing? value=null
        then: CALL 'public abstract fun invoke (p1: P1 of kotlin.Function1): R of kotlin.Function1 [operator] declared in kotlin.Function1' type=R of kotlin.Function1 origin=null
          $this: GET_VAR 'block: kotlin.Function1<kotlin.Any?, kotlin.Any?>? declared in com.example.ClassWithGenericsInConstructorMockff95e2a7ef5741dd9fe506074f570d58.<init>' type=kotlin.Function1<kotlin.Any?, kotlin.Any?>? origin=null
          p1: GET_VAR '<this>: com.example.ClassWithGenericsInConstructorMockff95e2a7ef5741dd9fe506074f570d58 declared in com.example.ClassWithGenericsInConstructorMockff95e2a7ef5741dd9fe506074f570d58' type=com.example.ClassWithGenericsInConstructorMockff95e2a7ef5741dd9fe506074f570d58 origin=null

	at org.jetbrains.kotlin.backend.jvm.codegen.FunctionCodegen.generate(FunctionCodegen.kt:47)
	at org.jetbrains.kotlin.backend.jvm.codegen.FunctionCodegen.generate$default(FunctionCodegen.kt:40)
	at org.jetbrains.kotlin.backend.jvm.codegen.ClassCodegen.generateMethodNode(ClassCodegen.kt:406)
	at org.jetbrains.kotlin.backend.jvm.codegen.ClassCodegen.generateMethod(ClassCodegen.kt:423)
	at org.jetbrains.kotlin.backend.jvm.codegen.ClassCodegen.generate(ClassCodegen.kt:168)
	at org.jetbrains.kotlin.backend.jvm.FileCodegen.lower(JvmPhases.kt:39)
	at org.jetbrains.kotlin.backend.common.phaser.PhaseFactoriesKt.createFilePhase$lambda$4(PhaseFactories.kt:71)
	at org.jetbrains.kotlin.backend.common.phaser.PhaseBuildersKt$createSimpleNamedCompilerPhase$1.phaseBody(PhaseBuilders.kt:69)
	at org.jetbrains.kotlin.backend.common.phaser.SimpleNamedCompilerPhase.phaseBody(CompilerPhase.kt:226)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedCompilerPhase.invoke(CompilerPhase.kt:113)
	at org.jetbrains.kotlin.backend.common.phaser.PerformByIrFilePhase.invokeSequential(performByIrFile.kt:62)
	... 45 more
Caused by: java.lang.AssertionError: Non-reified type parameter under ::class should be rejected by type checker: R of com.example.ClassWithGenericsInConstructor
	at org.jetbrains.kotlin.backend.jvm.codegen.ExpressionCodegen.generateClassLiteralReference(ExpressionCodegen.kt:1564)
	at org.jetbrains.kotlin.backend.jvm.codegen.ExpressionCodegen.visitClassReference(ExpressionCodegen.kt:1546)
	at org.jetbrains.kotlin.backend.jvm.codegen.ExpressionCodegen.visitClassReference(ExpressionCodegen.kt:138)
	at org.jetbrains.kotlin.ir.expressions.IrClassReference.accept(IrClassReference.kt:24)
	at org.jetbrains.kotlin.backend.jvm.codegen.ExpressionCodegen.gen(ExpressionCodegen.kt:217)
	at org.jetbrains.kotlin.backend.jvm.codegen.IrCallGenerator.genValueAndPut(IrCallGenerator.kt:48)
	at org.jetbrains.kotlin.backend.jvm.codegen.ExpressionCodegen.visitCall$handleParameter(ExpressionCodegen.kt:616)
	at org.jetbrains.kotlin.backend.jvm.codegen.ExpressionCodegen.visitCall(ExpressionCodegen.kt:644)
	at org.jetbrains.kotlin.backend.jvm.codegen.ExpressionCodegen.visitCall(ExpressionCodegen.kt:138)
	at org.jetbrains.kotlin.ir.expressions.IrCall.accept(IrCall.kt:24)
	at org.jetbrains.kotlin.backend.jvm.codegen.ExpressionCodegen.gen(ExpressionCodegen.kt:217)
	at org.jetbrains.kotlin.backend.jvm.codegen.ExpressionCodegen.generateConstructorArguments(ExpressionCodegen.kt:771)
	at org.jetbrains.kotlin.backend.jvm.codegen.ExpressionCodegen.visitDelegatingConstructorCall(ExpressionCodegen.kt:736)
	at org.jetbrains.kotlin.backend.jvm.codegen.ExpressionCodegen.visitDelegatingConstructorCall(ExpressionCodegen.kt:138)
	at org.jetbrains.kotlin.ir.expressions.IrDelegatingConstructorCall.accept(IrDelegatingConstructorCall.kt:21)
	at org.jetbrains.kotlin.backend.jvm.codegen.ExpressionCodegen.visitStatementContainer(ExpressionCodegen.kt:579)
	at org.jetbrains.kotlin.backend.jvm.codegen.ExpressionCodegen.visitBlockBody(ExpressionCodegen.kt:584)
	at org.jetbrains.kotlin.backend.jvm.codegen.ExpressionCodegen.visitBlockBody(ExpressionCodegen.kt:138)
	at org.jetbrains.kotlin.ir.expressions.IrBlockBody.accept(IrBlockBody.kt:20)
	at org.jetbrains.kotlin.backend.jvm.codegen.ExpressionCodegen.generate(ExpressionCodegen.kt:240)
	at org.jetbrains.kotlin.backend.jvm.codegen.FunctionCodegen.doGenerate(FunctionCodegen.kt:123)
	at org.jetbrains.kotlin.backend.jvm.codegen.FunctionCodegen.generate(FunctionCodegen.kt:44)
	... 55 more
```

**Comment by lupuuss** [comment]
Hi! Thank you for reporting this. Seems like a bug. I don't think there is a workaround for that. I'll try to fix it this week.

**Comment by lupuuss** [comment]
Fixed in [2.6.1]( https://github.com/lupuuss/Mokkery/releases/tag/v2.6.1)!
